logger.go

```diff
@@ -0,0 +1,175 @@
+package glog
+
+import (
+	"context"
+	"fmt"
+)
+
+// Logger provides logging functionality with additional data and prefixing.
+type Logger struct {
+	*loggingT
+	prefix string
+	// Extra arguments to be appended to each request
+	data []interface{}
+}
+
+// NewLogger creates a Logger instance with no additional data.
+func NewLogger() Logger {
+	return Logger{
+		loggingT: &logging,
+	}
+}
+
+// WithContext creates a logger from a context.Context
+func WithContext(ctx context.Context) Logger {
+	data, _ := ctx.Value(contextKeyData).([]interface{})
+	prefix, _ := ctx.Value(contextKeyPrefix).(string)
+	return Logger{
+		loggingT: &logging,
+		data:     data,
+		prefix:   prefix,
+	}
+}
+
+// WithPrefix creates a Logger with a given prefix.
+func WithPrefix(prefix string) Logger {
+	return Logger{
+		loggingT: &logging,
+		prefix:   prefix,
+	}
+}
+
+// WithData creates a Logger with a given set of backend data.
+// All args provided will be wrapped with glog.Data prior to being sent.
+func WithData(args ...interface{}) Logger {
+	return Logger{
+		loggingT: &logging,
+		data:     args,
+	}
+}
+
+// WithPrefix creates a Logger from an existing logger with a specified prefix.
+// Any prefix on the input Logger will be replaced.
+func (l Logger) WithPrefix(prefix string) Logger {
+	newLogger := l
```

Comment:
This will not copy the embedded pointer and the data. Is this intentional?

Reply:
Do you mean it will still reference the same underlying loggingT and data map? Yes, that's intentional, cutting down on a bunch of copied things that I don't understand ;)

Reply:
But then, why create a new one in the first place? Why not update the existing one?
The fact that some fields are copied and some are not is VERY confusing.

Reply:
(see reply above about why we create a new one).